### PR TITLE
feat!: api changes to `core` and `react` packages

### PR DESCRIPTION
**Pull request by dalechyn**

### CODE REVIEW

File: packages/core/src/types/utils.ts [hunk 1]
```diff
@@ -1,14 +1,99 @@
...
+  ? LooseOmit<type, keys>
+  : never
+
+export type UnionOmit<type, keys extends keyof type> = type extends any
+  ? Omit<type, keys>
+  : never
+
+export type UnionPartial<type> = type extends object ? ExactPartial<type> : type
+
+export type MaybeArray<T> = T | T[]
```
Comment (akaia-shadowfox) [inline packages/core/src/types/utils.ts:96]: This naming introduces some confusion, because `Maybe` is associated with the algebraic type

Comment (dalechyn) [reply packages/core/src/types/utils.ts:96]: Which kind of algebraic type does it associate with?

Comment (akaia-shadowfox) [reply packages/core/src/types/utils.ts:96]: `Maybe` aka `Option`

https://codinghelmet.com/articles/understanding-the-option-maybe-functional-type

https://en.wikipedia.org/wiki/Maybe_type

https://kevin-da-silva.medium.com/functional-programming-maybe-and-either-a1eb9db0d816#bc09

https://codewithstyle.info/advanced-functional-programming-in-typescript-maybe-monad/#Introducing-Maybe

The correct definition ( as long as we're not considering the true container type ) of `MaybeArray<T>` would be `T[] | null`, thus `MaybeArray<T> = T | T[]` is misleading.
It's essentially like saying that `Maybe<Array<T>> = T | T[]` or `Option<Array<T>> = T | T[]`


P.S. GraphQL codegen also [defines](https://github.com/dotansimha/graphql-code-generator/blob/b36509997251d1e3538634addbc52d0e87064113/examples/typescript-resolvers/src/type-defs.d.ts#L2) `Maybe<T>` as `T | null`

image

image

Comment (carina-akaia) [reply packages/core/src/types/utils.ts:96]: So, `StringOrArray` or something similar would be the proper fit here

Comment (dalechyn) [reply packages/core/src/types/utils.ts:96]: i think it's cool. https://github.com/search?q=MaybePromise+language:TypeScript+&type=code
https://github.com/search?q=MaybeArray+language:TypeScript+&type=code

you could check repos that use similar type naming.

Comment (akaia-shadowfox) [reply packages/core/src/types/utils.ts:96]: > you could check repos that use similar type naming.

That's the biggest collective mistake I've ever seen, but I give up on this one 🫡

**Review by akaia-shadowfox — APPROVED** [review]

**Pull request merged**